# Assignment 6: Data Wrangling


## Learning Objectives
This lesson meets the following learning objectives:
- The ability to use Python for data wrangling.

## Instructions
Read through all of the text in this page. This assignment provides step-by-step training divided into numbered sections. The sections often contain embeded exectable code for demonstration.  Section headers with icons have special meanings:  

- <i class="fas fa-puzzle-piece"></i> The puzzle icon indicates that the section provides a practice exercise that must be completed.  Follow the instructions for the exercise and do what it asks.  Exercises must be turned in for credit.
- <i class="fa fa-cogs"></i> The cogs icon indicates that the section provides a task to perform.  Follow the instructions to complete the task.  Tasks are not turned in for credit but must be completed to continue progress.

Review the list of items in the **Expected Outcomes** section to check that you feel comfortable with the material you just learned. If you do not, then take some time to re-review that material again. If after re-review you are not comfortable, do not feel confident or do not understand the material, please ask questions on Slack to help.

Follow the instructions in the **What to turn in** section to turn in the exercises of the assginment for course credit.

## Background
The purpose of this assignment is to build on Tidy data cleaning by using Python tools to "massage" or "wrangle" data into formats that are most useful for visualization and analytics.

**What is data wrangling?**

> Data wrangling, sometimes referred to as data munging, is the process of transforming and mapping data from one "raw" data form into another format with the intent of making it more appropriate and valuable for a variety of downstream purposes such as analytics. 

- [Data Wangling](https://en.wikipedia.org/wiki/Data_wrangling) *Wikipedia*

Previously, we learned about Tidy rules for reformatting data.  Transforming data into a Tidy dataset is data wrangling.  We have also learned to how to correct data types, remove missing values and duplicates.  This lessons is, therefore, an opportunity to bring everything together.  Some of the material will be a review, but should help reinforce the concepts.

## <i class="fa fa-cogs"></i> Notebook Setup
As before, we import any needed packages at the top of our notebook. Let's import Numpy and Pandas:

In [1]:
import numpy as np
import pandas as pd

## 1. Data Exploration
The first step in any data analytics task is import and exploration of data.  At this point, we have learned all of the steps we need to identify the data columns, their data types, recognize where we have missing values and recognize categorical and numeric variables in the data.   

For this tutorial we will use a dataset named "Abolone" from the [University of California Irvine Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/Abalone). The datafile is named `abalone.data` and is available in the data directory that accompanies this notebook.  The data has 10 "attributes" or variables. The following table describes these 10 variables, their types, and additional details.

<table>
    <tr><th>Name</th><th>Data Type</th><th>Metric</th><th>Description</th></tr>
    <tr><td>Sample ID</td><td>integer</td><td></td><td>A unique number for each sample taken</td></tr>
    <tr><td>Sex</td><td>nominal</td><td></td><td>M = 0, F = 1, and I = 2 (infant)</td></tr>
	<tr><td>Length</td><td>continuous</td><td>mm</td><td>Longest shell measurement</td></tr>
	<tr><td>Diameter</td><td>continuous</td><td>mm</td><td>perpendicular to length</td></tr>
	<tr><td>Height</td><td>continuous</td><td>mm</td><td>with meat in shell</td></tr>
	<tr><td>Whole weight</td><td>continuous</td><td>grams</td><td>whole abalone</td></tr>
	<tr><td>Shucked weight</td><td>continuous</td><td>grams</td><td>weight of meat</td></tr>
	<tr><td>Viscera weight</td><td>continuous</td><td>grams</td><td>gut weight (after bleeding)</td></tr>
	<tr><td>Shell weight</td><td>continuous</td><td>grams</td><td>after being dried</td></tr>
	<tr><td>Rings</td><td>integer</td><td></td><td>+1.5 gives the age in years</td></tr>
</table>

***Note:*** To demonstrate specific techniques of data wrangling, the dataset provided to you was altered: a sample ID column was added, the Sex column contains numeric IDs, and missing values were added as were duplicates.

This data has no header information, so, we'll provide it when we import the data:

In [2]:
abalone = pd.read_csv('data/abalone.data', header = None)
abalone.columns = ['Sample_ID','Sex', 'Length', 'Diameter', 'Height', 
          'Whole_weight', 'Shucked_weight', 'Viscera_weight', 
          'Shell_weight', 'Rings']
abalone.head()

Sample_ID  Sex  Length  Diameter  Height  Whole_weight  Shucked_weight  \
0       1000    0   0.455     0.365   0.095        0.5140          0.2245   
1       1001    0   0.350     0.265   0.090        0.2255          0.0995   
2       1002    1   0.530     0.420   0.135        0.6770          0.2565   
3       1003    0   0.440     0.365   0.125        0.5160          0.2155   
4       1004    2   0.330     0.255   0.080        0.2050          0.0895   

   Viscera_weight  Shell_weight  Rings  
0          0.1010         0.150     15  
1          0.0485         0.070      7  
2          0.1415         0.210      9  
3          0.1140         0.155     10  
4          0.0395         0.055      7

### 1.1 Exploring Data Types
First, let's explore how Pandas imported the data types:

In [3]:
abalone.dtypes

Sample_ID           int64
Sex                 int64
Length            float64
Diameter          float64
Height            float64
Whole_weight      float64
Shucked_weight    float64
Viscera_weight    float64
Shell_weight      float64
Rings               int64
dtype: object

Other than the first, second and last columns, all others were imported as `float64` which is a decimal value. The others were imported as an `integer`.  This looks correct for the data.

Let's get a sense of how big the data is:

In [4]:
abalone.shape

(4186, 10)

Next, we can explore the distribution of numerical data using the `describe` function:

In [4]:
abalone.describe()

Sample_ID          Sex       Length     Diameter       Height  \
count  4186.000000  4186.000000  4184.000000  4185.000000  4186.000000   
mean   3086.700669     0.955805     0.523635     0.407621     0.139406   
std    1206.816977     0.828239     0.120323     0.099446     0.041890   
min    1000.000000     0.000000     0.075000     0.055000     0.000000   
25%    2039.250000     0.000000     0.450000     0.350000     0.115000   
50%    3085.500000     1.000000     0.545000     0.425000     0.140000   
75%    4131.750000     2.000000     0.615000     0.480000     0.165000   
max    5178.000000     2.000000     0.815000     0.650000     1.130000   

       Whole_weight  Shucked_weight  Viscera_weight  Shell_weight        Rings  
count   4185.000000     4186.000000     4185.000000   4186.000000  4186.000000  
mean       0.827486        0.358964        0.180390      0.238535     9.930244  
std        0.490774        0.222266        0.109723      0.139352     3.222740  
min        0.002000        0.001000        0.000500      0.001500     1.000000  
25%        0.440500        0.185125        0.092500      0.130000     8.000000  
50%        0.798000        0.335500        0.170500      0.233000     9.000000  
75%        1.153000        0.501500        0.252500      0.328500    11.000000  
max        2.825500        1.488000        0.760000      1.005000    29.000000

Observe that even though the 'Sex' column was provided as a numeric value, it is actually meant to be categorical, with each sex represented as a unique number.  We can explore the categorical data using the `groupby` function, followed by the `size` function.

In [5]:
abalone.groupby(by=['Sex']).size()

Sex
0    1532
1    1307
2    1347
dtype: int64

### 1.2 Finding Missing Values
Before proceeding with any analysis you should know the state of missing values in the dataset.  For most analytics missing values are not supported. Some tools will automatically ignore them but it may be easier, in some cases, to remove them.

First, let's quantify how many missing values we have. The `isna` function will convert the data into `True` or `False` values: `True` if the value is missing:

In [6]:
abalone.isna().head()

Sample_ID    Sex  Length  Diameter  Height  Whole_weight  Shucked_weight  \
0      False  False   False     False   False         False           False   
1      False  False   False     False   False         False           False   
2      False  False   False     False   False         False           False   
3      False  False   False     False   False         False           False   
4      False  False   False     False   False         False           False   

   Viscera_weight  Shell_weight  Rings  
0           False         False  False  
1           False         False  False  
2           False         False  False  
3           False         False  False  
4           False         False  False

We can use the `sum` function to then identify how many missing values we have per column:

In [8]:
abalone.isna().sum()

Sample_ID         0
Sex               0
Length            2
Diameter          1
Height            0
Whole_weight      1
Shucked_weight    0
Viscera_weight    1
Shell_weight      0
Rings             0
dtype: int64

### 1.3 Inspecting Duplicates
Sometimes we may or may not want duplicates in the data. This depends on the expectations of the experiments and the measurements taken. Sometimes duplicates may represent human error in data entry. So, let's look for duplicated data.  We have 4,184 rows, let's see how many unique values per column that we have:

In [9]:
abalone.nunique()


Sample_ID         4179
Sex                  3
Length             134
Diameter           111
Height              51
Whole_weight      2429
Shucked_weight    1515
Viscera_weight     880
Shell_weight       926
Rings               28
dtype: int64

For all of the columns we have fewer that 4,184 values.  For columns like 'Sex' we have 3 unique values, but these repeated values are expected.  The decimal values also have duplicates. The likelihood of seeing the exact same decimal values varies based on the distribution for the variable and the number of decimal values in the measurement.  The number of duplicated values does not seem unordinary.  However, the sample ID should be unique, yet we have 4,177 of them instead of 4,184. This implies we have duplicated samples in the data. 

We can identify then umber of duplicated 'Sample_ID' values are in the data by using the `duplicated` function. 

In [10]:
abalone.duplicated(subset='Sample_ID').sum()

7

We have 7 duplicated rows. Now let's see which rows have duplicated samples:

In [10]:
abalone[abalone.duplicated(subset='Sample_ID', keep= False)]

Sample_ID  Sex  Length  Diameter  Height  Whole_weight  Shucked_weight  \
711        1711    0   0.375     0.300   0.100        0.2465          0.1040   
712        1712    2   0.280     0.205   0.055        0.1135          0.0450   
713        1713    0   0.355     0.265   0.085        0.2010          0.0690   
714        1714    0   0.350     0.255   0.080        0.1915          0.0800   
715        1715    2   0.275     0.200   0.065        0.1035          0.0475   
716        1716    2   0.290     0.205   0.070        0.0975          0.0360   
717        1717    2   0.250     0.190   0.060        0.0765          0.0360   
4179       1711    0   0.375     0.300   0.100        0.2465          0.1040   
4180       1712    2   0.280     0.205   0.055        0.1135          0.0450   
4181       1713    0   0.355     0.265   0.085        0.2010          0.0690   
4182       1714    0   0.350     0.255   0.080        0.1915          0.0800   
4183       1715    2   0.275     0.200   0.065        0.1035          0.0475   
4184       1716    2   0.290     0.205   0.070        0.0975          0.0360   
4185       1717    2   0.250     0.190   0.060        0.0765          0.0360   

      Viscera_weight  Shell_weight  Rings  
711           0.0475        0.0830     11  
712           0.0275        0.0335      7  
713           0.0530        0.0695      8  
714           0.0385        0.0630      9  
715           0.0205        0.0300      7  
716           0.0190        0.0350      8  
717           0.0115        0.0245      6  
4179          0.0475        0.0830     11  
4180          0.0275        0.0335      7  
4181          0.0530        0.0695      8  
4182          0.0385        0.0630      9  
4183          0.0205        0.0300      7  
4184          0.0190        0.0350      8  
4185          0.0115        0.0245      6

It looks like the rows are exact duplicates, so this was probably human entry error. We need to remove the copies rows. We will do so in the **3.1 Filtering** section below.

## 2. Cleanup
### 2.1 Correcting Data Types
During the data exploration phase above, we noticed that the Sex column was provided as a number to represent the Sex category, and therefore, Pandas imported that column as a numeric value. We need to convert that to a categorical value, because the meaning of the column is not ordinal or numeric. We should covert it to a string object.

We can do that with two functions that work on Series:  
- `astype`  converts the type of data in the series. 
- `replace`  replaces values in the series.

We'll use `astype` to convert the column to a string and `replace` to convert the numbers to more easily recognizable 'Male', 'Female' and 'Infant' strings.

In [11]:
# First convert the column from an integer to a string.
sex = abalone['Sex'].astype(str)

# Second convert 0 to Male, 1 to Female, and 2 to Infant.
sex = sex.replace('0', 'Male')
sex = sex.replace('1', 'Female')
sex = sex.replace('2', 'Infant')

# Now replace the 'Sex' column of the dataframe with the new Series.
abalone['Sex'] = sex
abalone.head()

Sample_ID     Sex  Length  Diameter  Height  Whole_weight  Shucked_weight  \
0       1000    Male   0.455     0.365   0.095        0.5140          0.2245   
1       1001    Male   0.350     0.265   0.090        0.2255          0.0995   
2       1002  Female   0.530     0.420   0.135        0.6770          0.2565   
3       1003    Male   0.440     0.365   0.125        0.5160          0.2155   
4       1004  Infant   0.330     0.255   0.080        0.2050          0.0895   

   Viscera_weight  Shell_weight  Rings  
0          0.1010         0.150     15  
1          0.0485         0.070      7  
2          0.1415         0.210      9  
3          0.1140         0.155     10  
4          0.0395         0.055      7

In addition, the Sample ID column, despite that it is numeric should not be treated as a numeric column, so let's convert that too:

In [12]:
# Convert Sample_ID to a string
abalone['Sample_ID'] = abalone['Sample_ID'].astype(str)

# Let's check out the datatypes to make sure they match our expectations:
abalone.dtypes

Sample_ID          object
Sex                object
Length            float64
Diameter          float64
Height            float64
Whole_weight      float64
Shucked_weight    float64
Viscera_weight    float64
Shell_weight      float64
Rings               int64
dtype: object

### 2.2 Handling Missing Values
As observed in section 2.2, we do indeed have missing values! Let's remove rows with missing values.  We can do so with the `dropna` function:

In [13]:
abalone = abalone.dropna(axis=0)
abalone.shape

(4184, 10)

Observe that the `axis` argument is set to 0 indicating we will remove rows with missing values. If we compare the `shape` of the dataframe now, with the shape when we first loaded it we will see that we have lost 2 rows with missing values.

In addition to `dropna` you can also use the `fillna` and `replace` functions to rewrite the missing values to something else.

### 2.3 Removing Duplicates

To remove duplicates we can use the [drop_duplicates](http://pandas.pydata.org/pandas-docs/version/0.17/generated/pandas.DataFrame.drop_duplicates.html) function of Pandas.  If we explore the duplicated columns of section 2.3 above we'll see that the rows are the same for all columns.  In this case we can call `drop_duplicates` with no arguments.  However, let's assume we can't guarantee that each column is the same,  but we do want to remove duplicated samples.  We can do this by using the `subset` argument of the `drop_duplicates` function. We don't want to drop all duplicates, we need to keep one set. Therefore, we'll use the `keep` argument to do this.

In [14]:
abalone = abalone.drop_duplicates(['Sample_ID'], keep='first')
abalone.shape

(4177, 10)

In practice, the `keep` argument will default to `first` so we don't need to provide it, but including it makes the code more clear.  We have now dropped all duplicated rows and we have 4,177 valid rows

## 3. Reshaping Data
Data reshaping is about altering the way data is housed in the data frames of Pandas. It includes filtering of rows, merging data frames, concatenating data frames, grouping, melting and pivoting. We have learned about all of these functions already. As a reminder, the following is a summary of what we've learned:

**Subsetting by Column**:
- *Indexing with column names*
  - Purpose: Allows you to slice the dataframe using column index names.
  - Introduced:  Pandas Part 1 Notebook
  - Example:
  ```python
   # Get the columns: Sample_ID, Sex, Height and Rings
   subset = abalone[['Sample_ID', 'Sex', 'Height', 'Rings']]
  ```
- *Indexing with the `loc` function*
  - Purpose: Allows you to slice the dataframe using row and column index names.
  - Introduced:  Pandas Part 1 Notebook
  - Example:
  ```python
   # Get the columns: Sample_ID, Sex, Height and Rings
   subset = abalone.loc[:,['Sample_ID', 'Sex', 'Height', 'Rings']]
  ```    
  
**Filtering Rows**:
- *Boolean Indexing*
  - Purpose: to filter rows that match desired criteria
  - Introduced:  Pandas Part 1 Notebook
  - Example: 
  
  ```python
   # Finds all rows with sex of "Male" and the number of rings > 3.
   matches = (abalone['Sex'] == 'Male') & (abalone['Rings'] > 3)
   male = abalone[matches]

   # Or more succinctly
   male = abalone[(abalone['Sex'] == 'Male') & (abalone['Rings'] > 3)]
  ```

**Grouping Data**:
- *`groupby` function*
  - Purpose:  To group rows together by "classes" or values of data. Allows you to perform aggregate functions, such as calculating means, summations, sizes, etc. You can create new data frames with aggregated values.
  - Introduced:  Pandas Part 2 Notebook. 
  - Example:
  ```python
  # Calculate the mean column value by each sex:
  abalone.groupby(by="Sex").mean()
  ```
  
**Merging DataFrames**:
- *`concat` function*
  - Purpose: To combine two dataframes.  Depending if the columns and row indexes are the same determines how the data frames are combined.
  - Introduced:  Pandas Part 2 Notebook.

**Melting**:
- *`melt` function*
  - Purpose:  Handles the case where categorical observations are stored in the header labels (i.e. violates Tidy rules).  It moves the header names into a new column and matches the corresponding values.
  - Introduced:  Tidy Part 1 Notebook.

**Pivoting**:
- *`pivot` and `pivot_table` functions*
  - Purpose: The opposite of `melt`. Uses unique values from one more columns to create new columns.
  - Intorduced: Tidy Part 1 Notebook.
  
You can use any of these functions/techniques to reshape the data to meet Tidy standards and appropriate for the analytic or visualization you want to perform.

## 4. <i class="fas fa-puzzle-piece"></i> Practice

Read in the apple data found in the `../data/apple_data.txt` directory.

In [23]:
apple = pd.read_csv("/Users/evandomsic/my_project/courses/Data-Analytics-With-Python/Module2-Data_Analytics/data/apple_data.tsv",sep='\t',na_values= ["--", "(D)"] )

Explore the data using the approach we've used thus far

In [24]:

print(apple.head)

print(apple.columns)

print(apple.dtypes)

print(apple.shape)

<bound method NDFrame.head of                 State    1982    1987    1992    1997    2002    2007  \
0             Alabama    899     626     474     479     265     231    
1              Alaska     NaN     NaN      2       4       9      10    
2             Arizona    214     224     235     158     185     496    
3            Arkansas    512     451     247     238     164     153    
4          California  2,904   2,787   3,082   3,454   2,120   2,074    
5            Colorado    608     595     542     493     361     468    
6         Connecticut    309     272     295     242     247     248    
7            Delaware     24      18      16      18      17      23    
8             Florida    196     149      92      81      53     126    
9             Georgia    893     669     525     428     305     286    
10             Hawaii     NaN     NaN     NaN     NaN      2       9    
11              Idaho    411     377     374     342     315     321    
12           Illinois

In [33]:
apple.describe()

State 1982 1987  1992  1997  2002 2007   1982   1987 1992.1  \
count            50   50   50    50    50    50   50     50     50     50   
unique           50   49   48    49    49    50   49     49     49     49   
top         Alabama   --   --  392   485   265   35      --     --    (D)   
freq              1    2    2     2     2     1    2      2      2      2   

        1997.1 2002.1 2007.1 2002.2 2007.2 2002.3 2007.3  
count       50     50     50     50     50     50     50  
unique      48     48     48     46     49     46     48  
top     5,389     (D)    (D)      8    427    (D)    (D)  
freq         2      3      3      2      2      5      3

In [35]:
apple.isna().sum()
#no NAS

       State    0
1982            0
1987            0
1992            0
1997            0
2002            0
2007            0
  1982          0
  1987          0
1992.1          0
1997.1          0
2002.1          0
2007.1          0
2002.2          0
2007.2          0
2002.3          0
2007.3          0
dtype: int64

In [20]:
#look for duplicates
apple.nunique()
apple.duplicated(subset="1982").sum()

apple

State    1982    1987    1992    1997    2002    2007  \
0             Alabama    899     626     474     479     265     231    
1              Alaska      --      --      2       4       9      10    
2             Arizona    214     224     235     158     185     496    
3            Arkansas    512     451     247     238     164     153    
4          California  2,904   2,787   3,082   3,454   2,120   2,074    
5            Colorado    608     595     542     493     361     468    
6         Connecticut    309     272     295     242     247     248    
7            Delaware     24      18      16      18      17      23    
8             Florida    196     149      92      81      53     126    
9             Georgia    893     669     525     428     305     286    
10             Hawaii      --      --      --      --      2       9    
11              Idaho    411     377     374     342     315     321    
12           Illinois    902     824     756     651     501     444    
13            Indiana    792     705     685     615     442     265    
14               Iowa    439     458     450     443     344     364    
15             Kansas    416     318     264     261     192     147    
16           Kentucky  1,022     916     878     710     616     623    
17          Louisiana    122     142      81      94      51      35    
18              Maine    400     386     388     349     374     345    
19           Maryland    551     453     392     338     288     252    
20      Massachusetts    452     526     477     399     338     369    
21           Michigan  2,788   2,516   2,473   2,290   1,750   1,772    
22          Minnesota    391     411     476     536     488     569    
23        Mississippi    413     293     258     248     159      80    
24           Missouri    871     673     539     524     485     438    
25            Montana    142     145     160     148     187     186    
26           Nebraska    100     124     126     106     107     125    
27             Nevada     50      56      45      47      45      35    
28      New Hampshire    219     209      227    210     183     215    
29         New Jersey    509     563     522     485     525     471    
30         New Mexico    740     666     748     806     907     807    
31           New York  2,055   1,726   1,578   1,557   1,447   1,350    
32     North Carolina  1,500   1,186     990     949     598     673    
33       North Dakota     38      29      32      28       5      21    
34               Ohio  1,857   1,614   1,454   1,404   1,282   1,000    
35           Oklahoma    695     582     392     427     239     178    
36             Oregon  2,053   1,784   1,684   1,763   1,485   1,330    
37       Pennsylvania  2,580   2,270   1,854   2,183   1,777   1,886    
38       Rhode Island     68      72      61      51      58      73    
39     South Carolina    390     321     330     292     339     205    
40       South Dakota     48      60      39      40      35      60    
41          Tennessee  1,321   1,065     924   1,003     357     543    
42              Texas    845     883     782     724     596     437    
43               Utah    635     568     536     485     388     457    
44            Vermont    243     217     251     237     243     264    
45           Virginia  1,400   1,160   1,057     970     746     626    
46         Washington  5,406   5,169   4,596   4,910   3,870   3,052    
47      West Virginia    712     586     505     593     463     494    
48          Wisconsin  1,027     856     965   1,006     886     935    
49            Wyoming     25      18      20      16      14      20    

        1982      1987    1992.1    1997.1    2002.1    2007.1 2002.2 2007.2  \
0     1,548     1,217     1,011       968       594       393     191    144   
1         --        --       (D)        2        13        15       8     10   
2     1,969     4,881     5,770     4,003     1,581 

Clean the data if needed

In [25]:
apple2 = apple.set_axis(['State', '1982_Farms', '1987_Farms', '1992_Farms', '1997_Farms', '2002_Farms', '2007_Farms', '1982_TTLAcres', '1987_TTLAcres', '1992_TTLAcres', '1997_TTLAcres', '2002_TTLAcres', '2007_TTLAcres', '2002_FarmBear_Acres', '2007_FarmBear_Acres', '2002_Bear_Acres', '2007_Bear_Acres' ], axis=1)
apple2.head()

State 1982_Farms 1987_Farms 1992_Farms 1997_Farms 2002_Farms  \
0        Alabama       899        626        474        479        265    
1         Alaska        NaN        NaN         2          4          9    
2        Arizona       214        224        235        158        185    
3       Arkansas       512        451        247        238        164    
4     California     2,904      2,787      3,082      3,454      2,120    

  2007_Farms 1982_TTLAcres 1987_TTLAcres 1992_TTLAcres 1997_TTLAcres  \
0       231         1,548         1,217         1,011           968    
1        10            NaN           NaN           NaN            2    
2       496         1,969         4,881         5,770         4,003    
3       153         2,276         2,120         1,430         1,048    
4     2,074        29,933        34,105        42,024        49,704    

  2002_TTLAcres 2007_TTLAcres 2002_FarmBear_Acres 2007_FarmBear_Acres  \
0          594           393                  191                 144   
1           13            15                    8                  10   
2        1,581         1,344                  118                 424   
3          658           287                  123                 116   
4       38,268        22,184                1,850               1,850   

  2002_Bear_Acres 2007_Bear_Acres  
0             386             307  
1             NaN             NaN  
2           1,459           1,249  
3             560             220  
4          35,856          20,954

In [26]:

#remove commas from numbers
apple2 = apple2.apply(lambda x: x.str.replace(',',''))

apple2


State 1982_Farms 1987_Farms 1992_Farms 1997_Farms 2002_Farms  \
0             Alabama       899        626        474        479        265    
1              Alaska        NaN        NaN         2          4          9    
2             Arizona       214        224        235        158        185    
3            Arkansas       512        451        247        238        164    
4          California      2904       2787       3082       3454       2120    
5            Colorado       608        595        542        493        361    
6         Connecticut       309        272        295        242        247    
7            Delaware        24         18         16         18         17    
8             Florida       196        149         92         81         53    
9             Georgia       893        669        525        428        305    
10             Hawaii        NaN        NaN        NaN        NaN         2    
11              Idaho       411        377        374        342        315    
12           Illinois       902        824        756        651        501    
13            Indiana       792        705        685        615        442    
14               Iowa       439        458        450        443        344    
15             Kansas       416        318        264        261        192    
16           Kentucky      1022        916        878        710        616    
17          Louisiana       122        142         81         94         51    
18              Maine       400        386        388        349        374    
19           Maryland       551        453        392        338        288    
20      Massachusetts       452        526        477        399        338    
21           Michigan      2788       2516       2473       2290       1750    
22          Minnesota       391        411        476        536        488    
23        Mississippi       413        293        258        248        159    
24           Missouri       871        673        539        524        485    
25            Montana       142        145        160        148        187    
26           Nebraska       100        124        126        106        107    
27             Nevada        50         56         45         47         45    
28      New Hampshire       219        209         227       210        183    
29         New Jersey       509        563        522        485        525    
30         New Mexico       740        666        748        806        907    
31           New York      2055       1726       1578       1557       1447    
32     North Carolina      1500       1186        990        949        598    
33       North Dakota        38         29         32         28          5    
34               Ohio      1857       1614       1454       1404       1282    
35           Oklahoma       695        582        392        427        239    
36             Oregon      2053       1784       1684       1763       1485    
37       Pennsylvania      2580       2270       1854       2183       1777    
38       Rhode Island        68         72         61         51         58    
39     South Carolina       390        321        330        292        339    
40       South Dakota        48         60         39         40         35    
41          Tennessee      1321       1065        924       1003        357    
42              Texas       845        883        782        724        596    
43               Utah       635        568        536        485        388    
44            Vermont       243        217        251        237        243    
45           Virginia      1400       1160       1057        970        746    
46         Washington      5406       5169       4596       4910       3870    
47      West Virginia       712        586        505        593        463    
48          Wisconsin      1027        856        965       1006        886    
49            Wy

In [28]:
apple2["State"] = apple2["State"]

apple2

State 1982_Farms 1987_Farms 1992_Farms 1997_Farms 2002_Farms  \
0             Alabama       899        626        474        479        265    
1              Alaska        NaN        NaN         2          4          9    
2             Arizona       214        224        235        158        185    
3            Arkansas       512        451        247        238        164    
4          California      2904       2787       3082       3454       2120    
5            Colorado       608        595        542        493        361    
6         Connecticut       309        272        295        242        247    
7            Delaware        24         18         16         18         17    
8             Florida       196        149         92         81         53    
9             Georgia       893        669        525        428        305    
10             Hawaii        NaN        NaN        NaN        NaN         2    
11              Idaho       411        377        374        342        315    
12           Illinois       902        824        756        651        501    
13            Indiana       792        705        685        615        442    
14               Iowa       439        458        450        443        344    
15             Kansas       416        318        264        261        192    
16           Kentucky      1022        916        878        710        616    
17          Louisiana       122        142         81         94         51    
18              Maine       400        386        388        349        374    
19           Maryland       551        453        392        338        288    
20      Massachusetts       452        526        477        399        338    
21           Michigan      2788       2516       2473       2290       1750    
22          Minnesota       391        411        476        536        488    
23        Mississippi       413        293        258        248        159    
24           Missouri       871        673        539        524        485    
25            Montana       142        145        160        148        187    
26           Nebraska       100        124        126        106        107    
27             Nevada        50         56         45         47         45    
28      New Hampshire       219        209         227       210        183    
29         New Jersey       509        563        522        485        525    
30         New Mexico       740        666        748        806        907    
31           New York      2055       1726       1578       1557       1447    
32     North Carolina      1500       1186        990        949        598    
33       North Dakota        38         29         32         28          5    
34               Ohio      1857       1614       1454       1404       1282    
35           Oklahoma       695        582        392        427        239    
36             Oregon      2053       1784       1684       1763       1485    
37       Pennsylvania      2580       2270       1854       2183       1777    
38       Rhode Island        68         72         61         51         58    
39     South Carolina       390        321        330        292        339    
40       South Dakota        48         60         39         40         35    
41          Tennessee      1321       1065        924       1003        357    
42              Texas       845        883        782        724        596    
43               Utah       635        568        536        485        388    
44            Vermont       243        217        251        237        243    
45           Virginia      1400       1160       1057        970        746    
46         Washington      5406       5169       4596       4910       3870    
47      West Virginia       712        586        505        593        463    
48          Wisconsin      1027        856        965       1006        886    
49            Wy

In [33]:
#change data type
apple2.iloc[:,1:18]= apple2.iloc[:,1:18].astype(float)

apple2.dtypes

State                   object
1982_Farms             float64
1987_Farms             float64
1992_Farms             float64
1997_Farms             float64
2002_Farms             float64
2007_Farms             float64
1982_TTLAcres          float64
1987_TTLAcres          float64
1992_TTLAcres          float64
1997_TTLAcres          float64
2002_TTLAcres          float64
2007_TTLAcres          float64
2002_FarmBear_Acres    float64
2007_FarmBear_Acres    float64
2002_Bear_Acres        float64
2007_Bear_Acres        float64
dtype: object

In [34]:
apple2.describe()

1982_Farms   1987_Farms   1992_Farms   1997_Farms   2002_Farms  \
count    48.00000    48.000000    49.000000    49.000000    50.000000   
mean    858.06250   764.958333   691.408163   690.510204   537.060000   
std     989.94688   915.886408   851.933974   912.523975   692.712221   
min      24.00000    18.000000     2.000000     4.000000     2.000000   
25%     237.00000   222.250000   235.000000   210.000000   168.750000   
50%     531.50000   544.500000   474.000000   428.000000   338.500000   
75%     932.00000   862.750000   782.000000   724.000000   578.250000   
max    5406.00000  5169.000000  4596.000000  4910.000000  3870.000000   

        2007_Farms  1982_TTLAcres  1987_TTLAcres  1992_TTLAcres  \
count    50.000000      48.000000      48.000000      47.000000   
mean    511.820000   12302.979167   12521.354167   12416.255319   
std     606.840183   25304.382357   27199.429306   28246.151550   
min       9.000000      41.000000      33.000000      35.000000   
25%     148.500000    1383.000000    1170.000000     999.000000   
50%     333.000000    4086.500000    4536.500000    4332.000000   
75%     562.500000    8642.000000    8791.250000    7407.000000   
max    3052.000000  145630.000000  162609.000000  169107.000000   

       1997_TTLAcres  2002_TTLAcres  2007_TTLAcres  2002_FarmBear_Acres  \
count      47.000000      47.000000      47.000000            50.000000   
mean    12927.510638    9863.234043    8470.638298           421.680000   
std     34141.187355   27100.575228   25504.573629           613.202348   
min         2.000000       6.000000      15.000000             2.000000   
25%       973.500000     596.500000     373.000000           119.250000   
50%      3187.000000    2384.000000    2057.000000           232.500000   
75%      6465.500000    4202.500000    3966.000000           400.750000   
max    215463.000000  172810.000000  165215.000000          3580.000000   

       2007_FarmBear_Acres  2002_Bear_Acres  2007_Bear_Acres  
count            50.000000        45.000000        47.000000  
mean            434.320000      8917.000000      7651.595745  
std             557.364954     24400.441667     23454.296203  
min               8.000000        19.000000         3.000000  
25%             116.500000       470.000000       261.500000  
50%             264.000000      2113.000000      1769.000000  
75%             453.250000      4083.000000      3444.000000  
max            2912.000000    152241.000000    152334.000000

Tidy the data

In [36]:
#Melt the data several times to generate appropriate data frames
apple_melt1 = pd.melt(apple2, id_vars='State', var_name='Year',
                      value_name='Farms', ignore_index=True)
apple_melt1.head(15)

State        Year   Farms
0          Alabama  1982_Farms   899.0
1           Alaska  1982_Farms     NaN
2          Arizona  1982_Farms   214.0
3         Arkansas  1982_Farms   512.0
4       California  1982_Farms  2904.0
5         Colorado  1982_Farms   608.0
6      Connecticut  1982_Farms   309.0
7         Delaware  1982_Farms    24.0
8          Florida  1982_Farms   196.0
9          Georgia  1982_Farms   893.0
10          Hawaii  1982_Farms     NaN
11           Idaho  1982_Farms   411.0
12        Illinois  1982_Farms   902.0
13         Indiana  1982_Farms   792.0
14            Iowa  1982_Farms   439.0

In [43]:
#Melt 2 
split_df = apple_melt1["Year"].str.split("_", n=1, expand=True)
apple_melt2 = pd.concat([apple_melt1, split_df], axis=1)
apple_melt2.head(10)
apple_melt2.columns = ["State", "Year_VarName", "Value", "Year", "VarName"]
apple_melt2.head(15)

State Year_VarName   Value  Year VarName
0          Alabama   1982_Farms   899.0  1982   Farms
1           Alaska   1982_Farms     NaN  1982   Farms
2          Arizona   1982_Farms   214.0  1982   Farms
3         Arkansas   1982_Farms   512.0  1982   Farms
4       California   1982_Farms  2904.0  1982   Farms
5         Colorado   1982_Farms   608.0  1982   Farms
6      Connecticut   1982_Farms   309.0  1982   Farms
7         Delaware   1982_Farms    24.0  1982   Farms
8          Florida   1982_Farms   196.0  1982   Farms
9          Georgia   1982_Farms   893.0  1982   Farms
10          Hawaii   1982_Farms     NaN  1982   Farms
11           Idaho   1982_Farms   411.0  1982   Farms
12        Illinois   1982_Farms   902.0  1982   Farms
13         Indiana   1982_Farms   792.0  1982   Farms
14            Iowa   1982_Farms   439.0  1982   Farms

In [44]:
#Drop unnecessary yeavarname

apple_melt2.drop("Year_VarName", inplace=True, axis=1)
apple_melt2.head(15)


State   Value  Year VarName
0          Alabama   899.0  1982   Farms
1           Alaska     NaN  1982   Farms
2          Arizona   214.0  1982   Farms
3         Arkansas   512.0  1982   Farms
4       California  2904.0  1982   Farms
5         Colorado   608.0  1982   Farms
6      Connecticut   309.0  1982   Farms
7         Delaware    24.0  1982   Farms
8          Florida   196.0  1982   Farms
9          Georgia   893.0  1982   Farms
10          Hawaii     NaN  1982   Farms
11           Idaho   411.0  1982   Farms
12        Illinois   902.0  1982   Farms
13         Indiana   792.0  1982   Farms
14            Iowa   439.0  1982   Farms

In [46]:
apple_tidy = pd.pivot(apple_melt2, index=["State", "Year"], columns = "VarName", values="Value")

apple_tidy.head(15)

VarName          Bear_Acres  FarmBear_Acres  Farms  TTLAcres
State      Year                                             
   Alabama 1982         NaN             NaN  899.0    1548.0
           1987         NaN             NaN  626.0    1217.0
           1992         NaN             NaN  474.0    1011.0
           1997         NaN             NaN  479.0     968.0
           2002       386.0           191.0  265.0     594.0
           2007       307.0           144.0  231.0     393.0
   Alaska  1982         NaN             NaN    NaN       NaN
           1987         NaN             NaN    NaN       NaN
           1992         NaN             NaN    2.0       NaN
           1997         NaN             NaN    4.0       2.0
           2002         NaN             8.0    9.0      13.0
           2007         NaN            10.0   10.0      15.0
   Arizona 1982         NaN             NaN  214.0    1969.0
           1987         NaN             NaN  224.0    4881.0
           1992         NaN             NaN  235.0    5770.0

In [48]:
#Remove index
apple_tidy.reset_index(inplace=True)
apple_tidy

VarName  index       State  Year  Bear_Acres  FarmBear_Acres  Farms  TTLAcres
0            0     Alabama  1982         NaN             NaN  899.0    1548.0
1            1     Alabama  1987         NaN             NaN  626.0    1217.0
2            2     Alabama  1992         NaN             NaN  474.0    1011.0
3            3     Alabama  1997         NaN             NaN  479.0     968.0
4            4     Alabama  2002       386.0           191.0  265.0     594.0
..         ...         ...   ...         ...             ...    ...       ...
295        295     Wyoming  1987         NaN             NaN   18.0      33.0
296        296     Wyoming  1992         NaN             NaN   20.0      53.0
297        297     Wyoming  1997         NaN             NaN   16.0      75.0
298        298     Wyoming  2002         NaN             8.0   14.0       NaN
299        299     Wyoming  2007        44.0            12.0   20.0      54.0

[300 rows x 7 columns]

## Expected Outcomes
At this point, you should feel comfortable with the following:
- How to begin exploration of any dataset prior to analysis
- Cleaning data and wrangling it into Tidy format.

## What to Turn in?
Be sure to **commit** and **push** your changes to this notebook.  All practice exercises should be completed.  Once completed, send a **Slack message** to the instructor indicating you have completed this assignment. The instructor will verify all work is completed. 